File: bindings/flutter/lib/src/model_package.dart
// Copyright (c) Microsoft Corporation. All rights reserved.
// Licensed under the MIT License.

import 'dart:convert';
import 'dart:ffi';

import 'package:ffi/ffi.dart';

import 'bindings/native_library.dart';
import 'error_capture.dart';
import 'model.dart';
import 'models/model_variant.dart';
import 'native_strings.dart';

/// A view of a [Model] that is known to be an ONNX Runtime model package.
///
/// Package handles delegate download / load / session state to their
/// **currently selected** variant. Call [selectBestVariant] first, or pin one
/// explicitly with [selectVariant].
class ModelPackage {
  ModelPackage.internal(this._model);

  final Model _model;

  /// Underlying model handle.
  Model get model => _model;

  int get _handle => _model.handle;

  /// The package manifest, scored against this device.
  ModelPackageManifest get manifest {
    final bindings = NativeLibrary.instance.bindings;
    final out = calloc<Pointer<Char>>();
    try {
      final status = bindings.flm_package_get_variants_json(_handle, out);
      checkStatus(status,
          fallbackMessage: 'flm_package_get_variants_json failed');
      try {
        final s = takeOutString(out.value);
        return ModelPackageManifest.fromJson(
            jsonDecode(s) as Map<String, Object?>);
      } finally {
        bindings.flm_string_free(out.value);
      }
    } finally {
      calloc.free(out);
    }
  }

  /// Convenience: variants as a list, scored against this device.
  List<ModelVariant> get variants => manifest.variants;

  /// Currently selected variant id, or null when none has been chosen.
  String? get selectedVariantId => manifest.selectedVariantId;

  /// Pin the package to a specific variant. Subsequent [Model.download] /
  /// [Model.load] calls on this package act on it.
  void selectVariant(String variantId) {
    withCString(variantId, (ptr) {
      final status =
          NativeLibrary.instance.bindings.flm_package_select_variant(_handle, ptr);
      checkStatus(status, fallbackMessage: 'flm_package_select_variant failed');
    });
  }

  /// Let the SDK pick the best variant for this device. Returns the id of the
  /// variant that was chosen.
  String selectBestVariant({VariantSelectionConstraints? constraints}) {
    final bindings = NativeLibrary.instance.bindings;
    final constraintsJson = constraints == null
        ? null
        : jsonEncode(constraints.toJson());
    final out = calloc<Pointer<Char>>();
    try {
      return withNullableCString(constraintsJson, (constraintsPtr) {
        final status = bindings.flm_package_select_best_variant(
            _handle, constraintsPtr, out);
        checkStatus(status,
            fallbackMessage: 'flm_package_select_best_variant failed');
        try {
          return takeOutString(out.value);
        } finally {
          bindings.flm_string_free(out.value);
        }
      });
    } finally {
      calloc.free(out);
    }
  }

  /// Obtain a standalone [Model] handle for one variant, so it can be
  /// downloaded / loaded / released independently of the package.
  Model getVariant(String variantId) {
    final bindings = NativeLibrary.instance.bindings;
    final out = calloc<Uint64>();
    try {
      return withCString(variantId, (ptr) {
        final status = bindings.flm_package_get_variant(_handle, ptr, out);
        checkStatus(status, fallbackMessage: 'flm_package_get_variant failed');
        return Model.fromHandle(out.value);
      });
    } finally {
      calloc.free(out);
    }
  }

  /// Estimate the transfer for a set of variants (or the currently selected
  /// variant when [variantIds] is null).
  DownloadEstimate estimateDownload({List<String>? variantIds}) {
    final bindings = NativeLibrary.instance.bindings;
    final json = variantIds == null ? null : jsonEncode(variantIds);
    final out = calloc<Pointer<Char>>();
    try {
      return withNullableCString(json, (jsonPtr) {
        final status = bindings.flm_package_estimate_download_json(
            _handle, jsonPtr, out);
        checkStatus(status,
            fallbackMessage: 'flm_package_estimate_download_json failed');
        try {
          final s = takeOutString(out.value);
          return DownloadEstimate.fromJson(
              jsonDecode(s) as Map<String, Object?>);
        } finally {
          bindings.flm_string_free(out.value);
        }
      });
    } finally {
      calloc.free(out);
    }
  }
}

File: bindings/ios/Sources/FoundryLocal/ModelPackage.swift
// Copyright (c) Microsoft Corporation. All rights reserved.
// Licensed under the MIT License.

import Foundation
import FoundryLocalCore

/// Model-package facet of ``Model``. On a non-package handle these calls throw
/// ``FoundryLocalError/Code/invalidState``.
extension Model {
    /// Enumerate this package's variants, each scored against the local device.
    ///
    /// `downloadSizeBytes` on the returned variants already excludes shared assets
    /// that happen to be present on disk — so it reflects what the user would
    /// actually transfer, not the theoretical variant size.
    public func variants() throws -> PackageVariants {
        let json = try readJSON { flm_package_get_variants_json(handle, $0) }
        return try flmJSONDecoder.decode(PackageVariants.self, from: Data(json.utf8))
    }

    /// Pin the package to a specific variant. Subsequent download / load / session
    /// calls act on it.
    public func selectVariant(_ variantId: String) throws {
        let status = variantId.withCString { flm_package_select_variant(handle, $0) }
        if status != FLM_OK {
            throw FoundryLocalError.fromCurrent(status: status)
        }
    }

    /// Let the SDK choose the best variant for this device, applying optional
    /// constraints. Returns the variant id that was chosen (also selected as the
    /// package's active variant).
    ///
    /// - Throws: ``FoundryLocalError/Code/incompatible`` when no variant satisfies the
    ///   constraints (e.g. an NPU-only constraint on a device with no NPU).
    @discardableResult
    public func selectBestVariant(_ constraints: VariantConstraints = .init()) throws -> String {
        let constraintsJSON = constraints.encodeAsJSON()
        var out: UnsafeMutablePointer<CChar>?
        let status = constraintsJSON.withCString { constraintsPtr in
            flm_package_select_best_variant(handle, constraintsPtr, &out)
        }
        if status != FLM_OK {
            throw FoundryLocalError.fromCurrent(status: status)
        }
        guard let ptr = out else {
            throw FoundryLocalError(code: .internalError, message: "select_best_variant returned no id")
        }
        defer { flm_string_free(ptr) }
        return String(cString: ptr)
    }

    /// Return a standalone handle for one variant, useful when an app wants to manage
    /// multiple variants side by side (e.g. downloading an NPU variant while a CPU
    /// variant serves requests).
    public func variant(_ variantId: String) throws -> Model {
        var out: flm_model = 0
        let status = variantId.withCString { flm_package_get_variant(handle, $0, &out) }
        if status != FLM_OK {
            throw FoundryLocalError.fromCurrent(status: status)
        }
        return Model(handle: out)
    }

    /// Estimate the transfer for a set of variants before committing to the download.
    /// Passing `nil` uses the currently selected variant.
    public func estimateDownload(variantIds: [String]? = nil) throws -> DownloadEstimate {
        let arg: String?
        if let variantIds {
            let data = try JSONSerialization.data(withJSONObject: variantIds, options: [.sortedKeys])
            arg = String(data: data, encoding: .utf8)
        } else {
            arg = nil
        }
        let json = try readJSON { out in
            if let arg {
                return arg.withCString { flm_package_estimate_download_json(handle, $0, out) }
            } else {
                return flm_package_estimate_download_json(handle, nil, out)
            }
        }
        return try flmJSONDecoder.decode(DownloadEstimate.self, from: Data(json.utf8))
    }
}

/// Constraints for ``Model/selectBestVariant``. Mirrors the JSON schema documented on
/// `flm_package_select_best_variant`.
public struct VariantConstraints: Sendable {
    /// Skip variants whose download would exceed this many bytes. `nil` = no limit.
    public var maxDownloadBytes: Int64?

    /// Only consider variants for these devices. `nil` = any device.
    public var allowedDevices: Set<Device>?

    /// Break ties on smallest download instead of highest compatibility score.
    public var preferSmallest: Bool = false

    public init(
        maxDownloadBytes: Int64? = nil,
        allowedDevices: Set<Device>? = nil,
        preferSmallest: Bool = false
    ) {
        self.maxDownloadBytes = maxDownloadBytes
        self.allowedDevices = allowedDevices
        self.preferSmallest = preferSmallest
    }

    func encodeAsJSON() -> String {
        var payload: [String: Any] = [:]
        if let maxDownloadBytes { payload["max_download_bytes"] = maxDownloadBytes }
        if let allowedDevices { payload["allowed_devices"] = allowedDevices.map(\.rawValue).sorted() }
        if preferSmallest { payload["prefer_smallest"] = true }
        return payload.jsonString() ?? "{}"
    }
}
